tests/cases/compiler/narrowByInstanceof.ts

```diff
@@ -0,0 +1,71 @@
+// @strict: true
+
+interface A { a: string }
+interface B { b: string }
+interface C { c: string }
+
+type AA = {
+    (): void;
+    prototype: A;
+}
+
+type BB = {
+    new(): B;
+}
+
+function foo(x: A | B | C, A: AA, B: BB, AB: AA | BB) {
+    if (x instanceof A) {

```

Comment:
I was really confused while reading this test because I was trying to understand how we weren't getting an error for trying to `instanceof` an interface, before I realized that `A` was being passed as a parameter and shadowed the outer declaration. If there's a different way to construct this that avoids that, that might be helpful for future reading / baseline reading.

Reply:
That is actually intended since this is precisely the "shadowing" that happens when you declare a class, i.e. for `class Foo {}` you end up with a value `Foo` (the constructor function) and a type `Foo` (the instance type). There isn't really any shadowing going on because the two namespaces are separate and determined by context.